Anonymous structs

AndreRenaud opened the pull request:
In the process of doing this there are a couple of other API adjustments as well:
- Moving from float32 -> float64. Phidgets API returns C.double, which is 64-bit
- Removing context & phidget from callbacks. This is required in C because otherwise there is no way of getting context into callbacks. But in Go we can use anonymous functions to capture context, so this feels more idiomatic to me

jrcichra commented:
This is fantastic! I assume I should merge this PR in first?

AndreRenaud commented:
Can you check it out and test it a bit at your end? I only have limited phidget devices here, essentially only a temperature sensor right now, so it's possible I've broken some things.

I'm also not 100% happy with losing the generic phidget interface and replacing it with a concrete struct. It means the demo program had to use []interface{}, which is a bit crappy. I'm wondering if Phidget should remain an interface, and have an unexpected struct that implements it instead.

jrcichra commented:
Sure. Let me add you to a private repo where I'm making use of this library for my home with lots of sensor types. If this program can work with all my sensors + this program it should be good to go.

My first attempt against my code shows some syntax changes are needed on my end with the current version, like you mentioned:
```
➜  read_phidgets2 git:(master) ✗ go build             
# github.com/jrcichra/read_phidgets2/readphidgets
readphidgets/sensor.go:74:13: cannot use &phidgets.PhidgetHumiditySensor{} (type *phidgets.PhidgetHumiditySensor) as type phidgets.Phidget in assignment
readphidgets/sensor.go:76:13: cannot use &phidgets.PhidgetTemperatureSensor{} (type *phidgets.PhidgetTemperatureSensor) as type phidgets.Phidget in assignment
readphidgets/sensor.go:78:13: cannot use &phidgets.PhidgetVoltageInput{} (type *phidgets.PhidgetVoltageInput) as type phidgets.Phidget in assignment
readphidgets/sensor.go:80:13: cannot use &phidgets.PhidgetCurrentInput{} (type *phidgets.PhidgetCurrentInput) as type phidgets.Phidget in assignment
readphidgets/sensor.go:82:13: cannot use &phidgets.PhidgetVoltageRatioInput{} (type *phidgets.PhidgetVoltageRatioInput) as type phidgets.Phidget in assignment
readphidgets/sensor.go:84:13: cannot use &phidgets.PhidgetLightSensor{} (type *phidgets.PhidgetLightSensor) as type phidgets.Phidget in assignment
readphidgets/sensor.go:86:13: cannot use &phidgets.PhidgetSoundSensor{} (type *phidgets.PhidgetSoundSensor) as type phidgets.Phidget in assignment
readphidgets/sensor.go:91:11: s.Phidget.Create undefined (type phidgets.Phidget has no field or method Create)
readphidgets/sensor.go:94:2: cannot type switch on non-interface value s.Phidget (type phidgets.Phidget)
readphidgets/sensor.go:96:4: p.SetIsHubPortDevice undefined (type *phidgets.PhidgetVoltageRatioInput has no field or method SetIsHubPortDevice)
readphidgets/sensor.go:96:4: too many errors
```

AndreRenaud commented:
I've redone it to still have a Phidget{} interface, and phidget (lowercase) is now the hidden internal implementation
This makes arrays of phidgets a bit nicer.
I added in the missing Phidget.SetIsHubPortDevice which I missed last time
Also dropped the unused ctx on SetonSPLChangeHandler

jrcichra commented on main.go:
```diff
@@ -32,16 +34,15 @@ func main() {
 		panic(err)
 	}
-	h.SetOnHumidityChangeHandler(func(p phidgets.Phidget, ctx interface{}, value float32) {
+	h.SetOnHumidityChangeHandler(func(value float64) {
```
Good point on removing the context concept for Go, since go has closures and C does not, having ctx just adds confusion.

jrcichra commented on main.go:
```diff
@@ -76,7 +76,8 @@ func main() {
 	for _, sensor := range sensors {
-		sensor.Close()
+		s := sensor.(io.Closer)
+		s.Close()
```
I'm not familiar what casting sensor to an (io.Closer) does...reading through here: https://golang.org/pkg/io/#Closer

Does it make sure I/O work is cleaned up within the sensor structure? In any case, can we somehow move this cast into `phidget.Phidgets` so it's not needed by library consumers?

AndreRenaud replied:
This was a hang-over from when I was using type Phidget struct {}, rather than an interface. It's no longer necessary, so I've reinstated the original behaviour.

jrcichra commented on phidgets/lcd.go:
```diff
@@ -10,148 +10,41 @@ void ccallback(void* handle, void* ctx, double b);  // Forward declaration.
-	cerr := C.Phidget_setHubPort(h, intToCInt(port))
+func (p *PhidgetLCD) SetText(text string) error {
+	str := C.CString(text)
+	cerr := C.PhidgetLCD_writeText(p.handle, C.FONT_6x12, 40, 25, str)
```
Note to self: I'd like to break out the `writeText` options in addition to this basic `SetText`. Better suited for a follow-up PR.

jrcichra commented on phidgets/phidget.go:
```diff
@@ -7,18 +7,126 @@ package phidgets
 //Phidget - general phidget interface that all phidgets are derived from (for ease of type management)
+type phidget struct {
+	handle C.PhidgetHandle
+	class  string
```
Any reason to break have `class` defined here vs. making the C call when necessary?

Here's my diff that runs:

```
diff --git a/phidgets/phidget.go b/phidgets/phidget.go
index 00e191e..11aa4a9 100644
--- a/phidgets/phidget.go
+++ b/phidgets/phidget.go
@@ -17,7 +17,6 @@ import (
 //Phidget - general phidget interface that all phidgets are derived from (for ease of type management)
 type phidget struct {
        handle C.PhidgetHandle
-       class  string
 }
 
 type Phidget interface {
@@ -46,7 +45,6 @@ func (p *phidget) OpenWaitForAttachment(timeout time.Duration) error {
        if cerr := C.Phidget_openWaitForAttachment(p.handle, C.uint(timeout.Milliseconds())); cerr != C.EPHIDGET_OK {
                return p.phidgetError(cerr)
        }
-       p.class, _ = p.GetChannelClassName()
        return nil
 }
 
@@ -56,8 +54,9 @@ func (p *phidget) phidgetError(cerr C.PhidgetReturnCode) error {
        }
        var errorString *C.char
        C.Phidget_getErrorDescription(cerr, &errorString)
-       if p.class != "" {
-               return errors.New(fmt.Sprintf("%s: %s", p.class, C.GoString(errorString)))
+       class, _ := p.GetChannelClassName()
+       if class != "" {
+               return errors.New(fmt.Sprintf("%s: %s", class, C.GoString(errorString)))
        }
        return errors.New(C.GoString(errorString))
 }
```

AndreRenaud replied:
There's a possible recursive issue there - p.GetChannelClassName called phidgetError. That was the reason to cache it. But we could just call the C.Phidget_getChannelClassName directly instead, that would be fine.

AndreRenaud replied:
How about like this?
```
diff --git a/phidgets/phidget.go b/phidgets/phidget.go
index 00e191e..a96cb37 100644
--- a/phidgets/phidget.go
+++ b/phidgets/phidget.go
@@ -46,7 +46,6 @@ func (p *phidget) OpenWaitForAttachment(timeout time.Duration) error {
        if cerr := C.Phidget_openWaitForAttachment(p.handle, C.uint(timeout.Milliseconds())); cerr != C.EPHIDGET_OK {
                return p.phidgetError(cerr)
        }
-       p.class, _ = p.GetChannelClassName()
        return nil
 }

@@ -55,9 +54,10 @@ func (p *phidget) phidgetError(cerr C.PhidgetReturnCode) error {
                return nil
        }
        var errorString *C.char
+       var className *C.char
        C.Phidget_getErrorDescription(cerr, &errorString)
-       if p.class != "" {
-               return errors.New(fmt.Sprintf("%s: %s", p.class, C.GoString(errorString)))
+       if C.Phidget_getChannelClassName(p.handle, &className) == C.EPHIDGET_OK {
+               return errors.New(fmt.Sprintf("%s: %s", C.GoString(className), C.GoString(errorString)))
        }
        return errors.New(C.GoString(errorString))
```

jrcichra replied:
Yeah, I guess the trade-off is making an extra C call on every phidgetError vs a small memory increase per phidget struct. The new method protects us in any weird situation where we get out of sync with the C lib. I'd prefer to keep state there.

AndreRenaud replied:
Yes, agreed. I would imagine that asking for the class of a phidget is actually just a constant table lookup anyway, so I doubt we're really saving anything by doing our own caching. Just the C.char* -> GoString, but that's also probably pretty tiny.

jrcichra replied:
I think you're right, I disassembled the latest branches and it looks like it's doing a `MOVQ` on x86_64 once we hit the C lib:

```
TEXT github.com/jrcichra/gophidgets/phidgets._Cfunc_Phidget_getChannelClassName(SB) _cgo_gotypes.go
  _cgo_gotypes.go:903	0x70cde0		64488b0c25f8ffffff	MOVQ FS:0xfffffff8, CX												
  _cgo_gotypes.go:903	0x70cde9		483b6110		CMPQ 0x10(CX), SP												
  _cgo_gotypes.go:903	0x70cded		0f8682000000		JBE 0x70ce75													
  _cgo_gotypes.go:903	0x70cdf3		4883ec28		SUBQ $0x28, SP													
  _cgo_gotypes.go:903	0x70cdf7		48896c2420		MOVQ BP, 0x20(SP)												
  _cgo_gotypes.go:903	0x70cdfc		488d6c2420		LEAQ 0x20(SP), BP												
  _cgo_gotypes.go:903	0x70ce01		c744244000000000	MOVL $0x0, 0x40(SP)												
  _cgo_gotypes.go:904	0x70ce09		488d442430		LEAQ 0x30(SP), AX												
  _cgo_gotypes.go:904	0x70ce0e		4889442418		MOVQ AX, 0x18(SP)												
  _cgo_gotypes.go:904	0x70ce13		488b0dee363600		MOVQ github.com/jrcichra/gophidgets/phidgets._cgo_7ad3368ddf45_Cfunc_Phidget_getChannelClassName(SB), CX	
  _cgo_gotypes.go:904	0x70ce1a		48890c24		MOVQ CX, 0(SP)													
  _cgo_gotypes.go:904	0x70ce1e		4889442408		MOVQ AX, 0x8(SP)												
  _cgo_gotypes.go:904	0x70ce23		e8d8b2cfff		CALL runtime.cgocall(SB)											
  _cgo_gotypes.go:905	0x70ce28		803d11d03d0000		CMPB $0x0, runtime.cgoAlwaysFalse(SB)										
  _cgo_gotypes.go:905	0x70ce2f		750a			JNE 0x70ce3b													
  _cgo_gotypes.go:909	0x70ce31		488b6c2420		MOVQ 0x20(SP), BP												
  _cgo_gotypes.go:909	0x70ce36		4883c428		ADDQ $0x28, SP													
  _cgo_gotypes.go:909	0x70ce3a		c3			RET														
  _cgo_gotypes.go:906	0x70ce3b		488b442430		MOVQ 0x30(SP), AX												
  _cgo_gotypes.go:906	0x70ce40		48890424		MOVQ AX, 0(SP)													
  _cgo_gotypes.go:906	0x70ce44		e8d72fd0ff		CALL runtime.convT64(SB)											
  _cgo_gotypes.go:906	0x70ce49		488d05109b0900		LEAQ 0x99b10(IP), AX												
  _cgo_gotypes.go:906	0x70ce50		48890424		MOVQ AX, 0(SP)													
  _cgo_gotypes.go:906	0x70ce54		e8c7aecfff		CALL runtime.cgoUse(SB)												
  _cgo_gotypes.go:907	0x70ce59		488d05405a0900		LEAQ 0x95a40(IP), AX												
  _cgo_gotypes.go:907	0x70ce60		48890424		MOVQ AX, 0(SP)													
  _cgo_gotypes.go:907	0x70ce64		488b442438		MOVQ 0x38(SP), AX												
  _cgo_gotypes.go:907	0x70ce69		4889442408		MOVQ AX, 0x8(SP)												
  _cgo_gotypes.go:907	0x70ce6e		e8adaecfff		CALL runtime.cgoUse(SB)												
  _cgo_gotypes.go:907	0x70ce73		ebbc			JMP 0x70ce31													
  _cgo_gotypes.go:903	0x70ce75		e8a672d6ff		CALL runtime.morestack_noctxt(SB)										
  _cgo_gotypes.go:903	0x70ce7a		e961ffffff		JMP github.com/jrcichra/gophidgets/phidgets._Cfunc_Phidget_getChannelClassName(SB)	
```

jrcichra replied:
Granted, I  could have opened up the phidget library (open source), but that's not as much fun :)

AndreRenaud commented:
If this all looks good, do you mind doing a merge/squash on it? Once it's there I'll go back and refactor the manager/discovery PR.
Thanks

jrcichra commented:
Merged! Thank you!
